go-controller/pkg/node/gateway_shared_intf.go

```diff
@@ -916,6 +924,94 @@ func (npw *nodePortWatcher) UpdateService(old, new *corev1.Service) error {
+			continue
+		}
+		for _, newExternalIP := range newExternalIPs {
+
+			newExternalIPPortTargetPorts.Insert(fmt.Sprintf("%s:%d:%s", newExternalIP, newPort.Port, newPort.TargetPort.String()))
+		}
+		for _, newClusterIP := range newClusterIPs {
+			newClusterIPPortTargetPorts.Insert(fmt.Sprintf("%s:%d:%s", newClusterIP, newPort.Port, newPort.TargetPort.String()))
+		}
+		for _, newNodeIP := range newNodeIPs {
+			newNodeIPPortTargetPorts.Insert(fmt.Sprintf("%s:%d:%s", newNodeIP.String(), newPort.NodePort, newPort.TargetPort.String()))
+		}
+	}
+
+	// Delete the conntrack entries for the old IP:port:targetPort combinations for External and LB IP, ClusterIP and NodePort
+	// that are not present in the corresponding new IP:port:targetPort combinations for External and LB IP, ClusterIP and NodePort
+	oldExternalIPs := util.GetExternalAndLBIPs(old)
+	oldClusterIPs := util.GetClusterIPs(old)
+	var oldNodeIPs []net.IP
+	if util.ServiceTypeHasNodePort(old) {
+		oldNodeIPs, _ = npw.nodeIPManager.ListAddresses()
+	}
+	deletedCount := 0
+	var errs []error
+	for _, oldPort := range old.Spec.Ports {
+		if oldPort.Protocol != corev1.ProtocolUDP {
+			continue
+		}
+		for _, oldExternalIP := range oldExternalIPs {
+			if !newExternalIPPortTargetPorts.Has(fmt.Sprintf("%s:%d:%s", oldExternalIP, oldPort.Port, oldPort.TargetPort.String())) {
+				if err := util.DeleteConntrackServicePort(oldExternalIP, oldPort.Port, oldPort.Protocol,
+					netlink.ConntrackOrigDstIP, nil); err != nil {
+					errs = append(errs, fmt.Errorf("failed to delete conntrack entry for service %s/%s with svcVIP %s, svcPort %d, protocol %s: %v",
+						old.Namespace, old.Name, oldExternalIP, oldPort.Port, oldPort.Protocol, err))
+				} else {
+					deletedCount++
+				}
+			}
+		}
+		for _, oldClusterIP := range oldClusterIPs {
+			if !newClusterIPPortTargetPorts.Has(fmt.Sprintf("%s:%d:%s", oldClusterIP, oldPort.Port, oldPort.TargetPort.String())) {
+				if err := util.DeleteConntrackServicePort(oldClusterIP, oldPort.Port, oldPort.Protocol,
+					netlink.ConntrackOrigDstIP, nil); err != nil {
+					errs = append(errs, fmt.Errorf("failed to delete conntrack entry for service %s/%s with svcVIP %s, svcPort %d, protocol %s: %v",
+						old.Namespace, old.Name, oldClusterIP, oldPort.Port, oldPort.Protocol, err))
+				} else {
+					deletedCount++
+				}
+			}
+		}
+		for _, oldNodeIP := range oldNodeIPs {
+			if !newNodeIPPortTargetPorts.Has(fmt.Sprintf("%s:%d:%s", oldNodeIP.String(), oldPort.NodePort, oldPort.TargetPort.String())) {
+				if err := util.DeleteConntrackServicePort(oldNodeIP.String(), oldPort.NodePort, oldPort.Protocol,
+					netlink.ConntrackOrigDstIP, nil); err != nil {
+					errs = append(errs, fmt.Errorf("failed to delete conntrack entry for service %s/%s with nodeIP %s, nodePort %d, protocol %s: %v",
+						old.Namespace, old.Name, oldNodeIP.String(), oldPort.NodePort, oldPort.Protocol, err))
+				} else {
+					deletedCount++
+				}
+			}
```

Comment:
The same pattern is repeated 3 times in the function (for externalIPs, clusterIPs, nodeIPs). This could be extracted into a helper function.